### DISCUSSION TITLE
Allow arbitrary boolean conditions in switch expressions

### DISCUSSION THREAD

**Discussion by olmobrutall** [dotnet/csharplang#5185] [Language Ideas] [answered]
Just to give context: I'm writing an algorithm that splits something an expression tree: 
```C#
Expression<Func<T, bool>> expression = a => a.FirstName == "John" && a.LastName = "Connor"
```
into a list of `(PropertyInfo pi, object? value)`, but this proposal has nothing to do with expression trees, so please don't tag it with [Blocked](https://github.com/dotnet/csharplang/issues/2545) yet :P 

Ideally I would write something like this: 

```C#
  var param = expression.Parameters.SingleEx();

  IEnumerable<(PropertyInfo pi, object? value)> PropPairs(Expression exp)
  {
      return exp switch 
      {
          BinaryExpression { NodeType: ExpressionType.And or ExpressionType.AndAlso } be => PropPairs(be.Left).Concat(PropPairs(be.Right)),
          BinaryExpression
          {
              NodeType: ExpressionType.Equal,
              Left: MemberExpression { Member: PropertyInfo pi, Expression: ParameterExpression p },
              Right: var other
          } && p.Equals(param) => new[] { (pi, ExpressionEvaluator.Eval(other)) },
          BinaryExpression
          {
              NodeType: ExpressionType.Equal,
              Left: var other,
              Right: MemberExpression { Member: PropertyInfo pi, Expression: ParameterExpression p },
          } && p.Equals(param) => new[] { (pi, ExpressionEvaluator.Eval(other)) }
          _ => throw new UnexpectedValueException(exp);
      };
}
```

But this is not allowed because:

* `&& p.Equals(param)` <-- CS1525	Invalid expression term '&&'	
* `and p == param`  <-- CS0029	Cannot implicitly convert type 'bool' to 'System.Linq.Expressions.BinaryExpression'
* `and p is param ` <-- CS0150	A constant value is expected

So I get it... switch expression only want patterns, not boolean conditions. But why?

Note that I don't want to put the condition after the `=>` because in examples like the following the second switch branch should fail and should try with the third one: 

```C#
Expression<Func<T, bool>> expression = a => DateTime.Now == a.DateOfBirth
```

So my proposal is to **relax switch expression to allow arbitrary boolean conditions**. I suppose this will complicate/disable exhaustive checking in some cases but I think is worth because... you get a warning anyway in most (every?) case if you don't have a discard pattern. 

## Current Alternative

You can write if with `if`/`is` but then you have to rename all the declared variables to avoid conflicts. 

 ```C#
  var param = expression.Parameters.SingleEx();
  IEnumerable<(PropertyInfo pi, object? value)> PropPairs(Expression exp)
  {
      if (exp is BinaryExpression { NodeType: ExpressionType.And or ExpressionType.AndAlso } be)
          return PropPairs(be.Left).Concat(PropPairs(be.Right));

      if (exp is BinaryExpression
          {
              NodeType: ExpressionType.Equal,
              Left: MemberExpression { Member: PropertyInfo pi, Expression: ParameterExpression p },
              Right: var other
          } && p.Equals(param))
          return new[] { (pi, ExpressionEvaluator.Eval(other)) };

      if (exp is BinaryExpression
          {
              NodeType: ExpressionType.Equal,
              Left: var other2,
              Right: MemberExpression { Member: PropertyInfo pi2, Expression: ParameterExpression p2 },
          } && p2.Equals(param))
          return new[] { (pi2, ExpressionEvaluator.Eval(other2)) };

      throw new UnexpectedValueException(exp);
  }
```

I would like to use switch expressions more often, but I often have to fall back to `if`/`is` solutions.

**Comment by ufcpp** [comment, marked as answer]
Can't you use `when` instead of `&&`?

**Reply by olmobrutall** [reply]
You are right. I overlooked it, but have now found it in the documentation: 

For switch expression: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression#case-guards
For switch statement: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/statements/selection-statements#case-guards

Thanks!
